frontend/lib/resolution/Resolver.cpp

```diff
@@ -4870,6 +4870,59 @@ void Resolver::exit(const Range* range) {
   }
 }
 
+bool Resolver::enter(const uast::Array* decl) {
+  return true;
+}
+void Resolver::exit(const uast::Array* decl) {
+  if (scopeResolveOnly) {
+    return;
+  }
+
+  ResolvedExpression& r = byPostorder.byAst(decl);
+
+  // Resolve call to appropriate array builder proc
+  const char* arrayBuilderProc;
+  std::vector<CallInfoActual> actuals;
+  if (!decl->isMultiDim()) {
+    arrayBuilderProc = "chpl__buildArrayExpr";
+  } else {
+    arrayBuilderProc = "chpl__buildNDArrayExpr";
+
+    // Get shape arg
+    std::vector<QualifiedType> shapeTupleElts;
+    for (auto dim : decl->shape()) {
+      shapeTupleElts.push_back(QualifiedType::makeParamInt(context, dim));
+    }
+    auto shapeTupleType = TupleType::getQualifiedTuple(context, shapeTupleElts);
+    actuals.emplace_back(
+        QualifiedType(QualifiedType::CONST_VAR, shapeTupleType),
+        UniqueString());
+  }
+
+  // Add element args
+  for (auto expr : decl->flattenedExprs()) {
+    auto exprType = byPostorder.byAst(expr).type();
+    // Short circuit if any elements have unknown type, since we won't be
+    // able to resolve the array builder proc.
+    if (exprType.isUnknown()) {
+      r.setType(QualifiedType());
+      return;
+    }
+    actuals.emplace_back(exprType, UniqueString());
+  }
+
+  auto ci = CallInfo(/* name */ UniqueString::get(context, arrayBuilderProc),
+                     /* calledType */ QualifiedType(),
+                     /* isMethodCall */ false,
+                     /* hasQuestionArg */ false,
+                     /* isParenless */ false, actuals);
+  auto scope = currentScope();
+  auto inScopes = CallScopeInfo::forNormalCall(scope, poiScope);
+  auto c = resolveGeneratedCall(decl, &ci, &inScopes);
```

Comment:
you might want to make a call to `shouldSkipCallResolution` (or whatever it's called). Particularly, I see lots of fun cases:

```Chapel
record R {
  var x;
  var y;
  var A = [x, y];
}
```

During initial field resolution, x and y will both be unknown or AnyType, and you can get into trouble passing uninstantiated types to the builder function.

Similarly, in dependently typed functions:

```Chapel
proc foo(x, y: something(x.type), A = [y]) {}
```

Above, when computing the initial type signature, `y` will have `Unknown` type. It's not good trying to resolve calls with unknown elements.

Reply:
yeah, `nullptr` is an acceptable choice. You may hit a segfault because there's a missing null check though (I ran into it on my WIP branch recently), but it's an easy fix